File: cits3001/labs/lab4/java/Lab4.java
package lab4.java;

import java.util.*;

public class Lab4 {

    public static void main(String[] args) {
        Lab4.test();
    }


    public static void test() {
        String[] dictionary = new String[] {"AIM", "ARM", "ART", "RIM", "RAM", "RAT", "ROT", "RUM", "RUN", "BOT", "JAM", "JOB", "JAB", "LAB", "LOB", "LOG", "SUN"};

        String[] solution = Lab4.findPath(dictionary, "AIM", "BOT");

        for (String word : solution) {
            System.out.println(word);
        }
    }


    /**
     * Finds a shortest sequence of words in the dictionary such that the first word is the startWord, 
    * the last word is the endWord, and each word is equal to the previous word with one letter changed.
    * All words in the sequence are the same length. If no sequence is possible, an empty array is returned.
    * It is assumed that both startWord and endWord are elements of the dictionary.
    * @param dictionary The set of words that can be used in the sequence; all words in the dictionary are capitalised.
    * @param startWord the first word on the sequence.
    * @param endWord the last word in the sequence.
    * @return an array containing a shortest sequence from startWord to endWord, in order, 
    * using only words from the dictionary that differ by a single character.
    * */
    public static String[] findPath(String[] dictionary, String startWord, String endWord){

        // convert dictionary to a set
        Set<String> dictSet = new HashSet<String>(Arrays.asList(dictionary));

        // Keep track of neighbours found for each node we visit
        HashMap<String, ArrayList<String>> nodeNeighbours = new HashMap<String, ArrayList<String>>();

        // Keep track of how far away a node is from the start word
        HashMap<String, Integer> distances = new HashMap<String, Integer>();

        // List to store our solution in 
        ArrayList<String> solution = new ArrayList<String>();


        // Perform a BFS to get nodes and their distances from START. 
        // Do this until we find the end node.

        // add each word in dict to our node neighbours map
        for (String node : dictSet) {
            nodeNeighbours.put(node, new ArrayList<String>());
        }

        Queue<String> queue = new LinkedList<String>();
        queue.offer(startWord);
        distances.put(startWord, 0);


        while (!queue.isEmpty()) {
            int count = queue.size();
            boolean foundEnd = false;

            for (int i = 0; i < count; i++) {
                String currNode = queue.poll();

                int currNodeDist = distances.get(currNode);

                ArrayList<String> currNodeNeighbours = getNeighbours(currNode, dictSet);

                for (String neighbour : currNodeNeighbours) {
                    nodeNeighbours.get(currNode).add(neighbour);

                    if (!distances.containsKey(neighbour)) {
                        distances.put(neighbour, currNodeDist + 1);

                        if (!endWord.equals(neighbour)) {
                            queue.offer(neighbour);
                        }
                        else {
                            foundEnd = true;
                            break;
                        }
                    }
                }
                if (foundEnd) {break;}
            }
            if (foundEnd) {break;}
        }

        // if no path found return an empty array
        if (!distances.containsKey(endWord)) {
            return new String[0];
        }

        getPath(startWord, endWord, nodeNeighbours, distances, solution);


        return solution.toArray(new String[0]);
    }


    private static boolean getPath(String currNode, 
                                String endNode, 
                                HashMap<String, ArrayList<String>> nodeNeighbours, 
                                HashMap<String, Integer> distances,
                                ArrayList<String> solution) 
    {
        int solutionDist = distances.get(endNode) + 1;

        solution.add(currNode);

        if (solution.size() == solutionDist) {
            if (currNode.equals(endNode)) {
                return true;
            }
            else {
                return false;
            }
        }

        for (String neighbour : nodeNeighbours.get(currNode)) {
            if (distances.get(currNode) + 1 == distances.get(neighbour)) {
                if (!getPath(neighbour, endNode, nodeNeighbours, distances, solution)) {
                    solution.remove(solution.size()-1);
                }
                else {
                    return true;
                }
            }
        }
        return false; // this should never happen
    }


    private static ArrayList<String> getNeighbours(String node, Set<String> dict) {
        ArrayList<String> neighbours  = new ArrayList<String>();
        char[] node_chars = node.toCharArray();

        for (char ch = 'A'; ch <= 'Z'; ch++) {
            for (int i = 0; i < node_chars.length; i++) {
                if (node_chars[i] != ch) {
                    char old_ch = node_chars[i]; // save old char to put back after checking dict
                    node_chars[i] = ch;
                    if (dict.contains(String.valueOf(node_chars))) {
                        neighbours.add(String.valueOf(node_chars));
                    }
                    node_chars[i] = old_ch;
                }
            }
        }
        return neighbours;
    }
}

File: cits3001/labs/lab4/python/lab4.py
""" CITS3001 Lab 4

Thomas Cleary 21704985
"""

import lab4tests

from collections import deque


def find_path(dictionary, start_word, end_word):
    '''
    returns a list of the word in the shortest path
    from start_word to end_word,
    where successive words are different in only one letter.

    types: dictionary (list), start/end_word (str)
    '''
    dictionary_set = set(dictionary)

    node_neighbours = {key : [] for key in dictionary_set}
    distances       = {start_word : 0}

    queue = deque() # infinite queue
    queue.append(start_word)

    while len(queue) > 0:
        count = len(queue)
        found_end = False

        for i in range(0, count):
            curr_node = queue.popleft()
            curr_node_dist = distances[curr_node]
            curr_node_neighbours = get_neighbours(curr_node, dictionary_set)

            for neighbour in curr_node_neighbours:
                node_neighbours[curr_node].append(neighbour)

                if neighbour not in distances.keys():  # if we have not already visited this neighbour
                    distances[neighbour] = curr_node_dist + 1

                    if neighbour != end_word:
                        queue.append(neighbour)

                    # found the shortest path
                    else: 
                        found_end = True
                        break
            if found_end:
                break
        if found_end:
            break

    # no path found, return an empty list
    if end_word not in distances.keys():
        return []
    
    path = []
    get_path(start_word, end_word, node_neighbours, distances, path)

    return path

 
def get_path(curr_node, end_node, node_neighbours, distances, solution):
    """ recursively build the path into the solution list """
    solution_dist = distances[end_node] + 1

    solution.append(curr_node)

    if len(solution) == solution_dist:
        if curr_node == end_node:
            return True
        else:
            return False

    for neighbour in node_neighbours[curr_node]:

        if distances[curr_node] + 1 == distances[neighbour]:
            if not get_path(neighbour, end_node, node_neighbours, distances, solution):
                del solution[-1]
            else:
                return True



def get_neighbours(node, dictionary):
    """ return a list of all neighbours of the node """
    neighbours = []

    # for characters a->z
    for ordinance in range(ord("A"), ord("Z") + 1):
        char = chr(ordinance)
        node_chars = list(node)

        for i, node_char in enumerate(node_chars):
            if node_char != char:
                neighbour = [char for char in node_chars] # deep copy
                neighbour[i] = char
                neighbour = "".join(neighbour)

                if neighbour in dictionary:
                    neighbours.append(neighbour)

    return neighbours


def get_dictionary(filename):
    """ return the dictionary of words as a list """
    with open(filename, "r") as dict_file:
        dictionary = []
        for line in dict_file:
            dictionary.append(line.strip())

    return dictionary


def main():
    """ run tests """
    small_dictionary = ['AIM', 'ARM', 'ART', 'RIM', 'RAM', 'RAT', 'ROT', 'RUM', 'RUN', 'BOT', 'JAM', 'JOB', 'JAB', 'LAB', 'LOB', 'LOG', 'SUN']
    large_dictionary = get_dictionary("../Dictionary.txt")


    lab4tests.lab4_test("Basic Test", 1, "AIM", "BOT", 6, small_dictionary)
    lab4tests.lab4_test("Basic Test", 2, "LOG", "JAM", 5, small_dictionary)
    lab4tests.lab4_test("Basic Test", 3, "LOG", "SUN", 9, small_dictionary)

    # test for start and end with no path
    lab4tests.lab4_test("No Solution Test", 1, "AIM", "ZZZ", 0, ["AIM", "ARM", "ART", "ZZZ"])


if __name__ == "__main__":
    main()
